batch/sql/estimated-current.sql

```diff
@@ -1231,6 +1266,22 @@ BEGIN
               jobs.job_id < cur_update_start_job_id + staging_n_jobs;
       END IF;
 
+      INSERT INTO job_group_self_and_ancestors (batch_id, job_group_id, ancestor_id, level)
+      SELECT new_ancestors.batch_id, new_ancestors.job_group_id, new_ancestors.ancestor_id, new_ancestors.level
+      FROM (
+        SELECT batch_id, job_group_id, MIN(ancestor_id) AS last_known_ancestor, MAX(level) AS last_known_level
```

Comment:
it's unclear to me what "last" means. Is this closest to the leaf or the root?

Reply:
This would be from walking up the tree starting at the leaf asking what is the last known parent, grandparent, great-grandparent etc. If we reach the root job group id in this query, there's nothing to fill in. Otherwise, we need to fill in the parents of the last known ancestor for that job.

Unfortunately, this complexity is needed because we need to support adding job groups in parallel just like we do with jobs and there's no guarantee the job groups will be inserted in the same chunk as it's parents and in the same order.

Reply:
> we need to support adding job groups in parallel just like we do with jobs

How come? We control the client here, how many job groups would someone reasonably be making in a single request? It doesn't seem unreasonable to me if someone has many little groups (which seems like an anti-pattern given the discussions we were having) to just make those up front and then submit jobs (which presumably are many more than the groups they have) in parallel.

Reply:
Siwei's use case would have benefitted from making 30K job groups. One for each transcript or gene in the genome.

Reply:
Ah ok, and those were arbitrarily nested? I think what's odd to me is not the creation of job groups in parallel but the fact that a parent job group might not exist prior to the addition of a child job group. I'm having a hard time imagining such a deep tree all created up front that it would be a performance requirement to be able to add a child before its parent.

Reply:
Think about how jobs are created. We do that in 6-way parallelism. There's no guarantee on the order of the API calls.

```python3
        await bounded_gather(
            *[functools.partial(self._submit_jobs, update_id, bunch, size, progress_task)
              for bunch, size in zip(byte_job_specs_bunches, bunch_sizes)
              ],
            parallelism=6,
            cancel_on_error=True,
        )
```